src/apphosting/githubConnections.ts

```diff
@@ -257,6 +259,32 @@ async function promptCloneUri(
   return { cloneUri, connection: cloneUriToConnection[cloneUri] };
 }
 
+/**
+ * Prompts the user for a GitHub branch and validates that the given branch
+ * actually exists. User is re-prompted until they enter a valid branch.
+ */
+export async function promptGitHubBranch(repoLink: devConnect.GitRepositoryLink) {
+  const { lookupMap } = await devConnect.listAllBranches(repoLink.name);
+
+  let branch = DEFAULT_GITHUB_BRANCH;
+  do {
+    branch = await promptOnce({
+      name: "branch",
+      type: "input",
+      default: DEFAULT_GITHUB_BRANCH,
+      message: "Pick a branch for continuous deployment",
+    });
+
+    if (!lookupMap.has(branch)) {
+      utils.logWarning(
+        "the given branch does not exist, please enter a valid branch for your repo",
+      );
+    }
+  } while (!lookupMap.has(branch));
```

Comment:
very optional suggestion, if you want to change this to use positive booleans instead (go/tott/685):

```
while true:
  branch = ...
  if (map.has(branch)) return branch
  utils.logWarning(...)
```